### ISSUE TITLE
Cannot access MongoDB with secrets in the connection string

### ISSUE THREAD

**Issue by mhdirkse** [ibissource/iaf#4526]
**Describe the issue**
To access MongoDB, I needed to put a `<Resource>` element with the connection string in `context.xml`. I could access my MongoDB database by adding my credentials, but this is not secure because `context.xml` is checked in. I modified my project to reference secrets, but then the Tomcat window produced an exception:

```
17-Feb-2023 16:15:50.956 INFO [main] org.apache.catalina.core.ApplicationContext.log Starting IbisContext
17-Feb-2023 16:15:56.567 WARNING [starting Adapter QueryMongoDB] org.apache.naming.NamingContext.lookup Unexpected exception resolving reference
        java.lang.IllegalArgumentException: state should be: databaseName does not contain '/'
                at com.mongodb.MongoNamespace.checkDatabaseNameValidity(MongoNamespace.java:65)
                at com.mongodb.ConnectionString.<init>(ConnectionString.java:390)
                at com.mongodb.client.MongoClients.create(MongoClients.java:61)
                at com.mongodb.client.MongoClientFactory.getObjectInstance(MongoClientFactory.java:97)
                at org.apache.naming.factory.FactoryBase.getObjectInstance(FactoryBase.java:96)
                at javax.naming.spi.NamingManager.getObjectInstance(NamingManager.java:332)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:864)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:158)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:850)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:158)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:850)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:158)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:850)
                at org.apache.naming.NamingContext.lookup(NamingContext.java:172)
                at org.apache.naming.SelectorContext.lookup(SelectorContext.java:161)
                at javax.naming.InitialContext.lookup(InitialContext.java:417)
                at org.springframework.jndi.JndiTemplate.lambda$lookup$0(JndiTemplate.java:157)
                at org.springframework.jndi.JndiTemplate.execute(JndiTemplate.java:92)
                at org.springframework.jndi.JndiTemplate.lookup(JndiTemplate.java:157)
                at org.springframework.jndi.JndiTemplate.lookup(JndiTemplate.java:179)
                at org.springframework.jndi.JndiLocatorSupport.lookup(JndiLocatorSupport.java:96)
                at nl.nn.adapterframework.jndi.ObjectLocator.lookup(ObjectLocator.java:40)
                at nl.nn.adapterframework.jndi.ObjectLocator.lookup(ObjectLocator.java:47)
                at nl.nn.adapterframework.jndi.JndiObjectFactory.lookup(JndiObjectFactory.java:57)
                at nl.nn.adapterframework.jndi.ObjectFactoryBase.compute(ObjectFactoryBase.java:66)
                at nl.nn.adapterframework.jndi.ObjectFactoryBase.lambda$get$0(ObjectFactoryBase.java:61)
                at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
                at nl.nn.adapterframework.jndi.ObjectFactoryBase.get(ObjectFactoryBase.java:61)
                at nl.nn.adapterframework.jndi.ObjectFactoryBase.get(ObjectFactoryBase.java:57)
                at nl.nn.adapterframework.mongodb.JndiMongoClientFactory.getMongoClient(JndiMongoClientFactory.java:46)
                at nl.nn.adapterframework.mongodb.MongoDbSender.open(MongoDbSender.java:149)
                at nl.nn.adapterframework.mongodb.MongoDbSender$$FastClassBySpringCGLIB$$c223776a.invoke(<generated>)
                at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
                at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
                at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
                at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
                at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
                at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
                at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
                at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
                at nl.nn.adapterframework.mongodb.MongoDbSender$$EnhancerBySpringCGLIB$$937b9510.open(<generated>)
                at nl.nn.adapterframework.pipes.MessageSendingPipe.start(MessageSendingPipe.java:854)
                at nl.nn.adapterframework.pipes.MessageSendingPipe$$FastClassBySpringCGLIB$$b131b2cb.invoke(<generated>)
                at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
                at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:793)
                at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
                at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
                at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:97)
                at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
                at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:763)
                at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:708)
                at nl.nn.adapterframework.pipes.SenderPipe$$EnhancerBySpringCGLIB$$7d266630.start(<generated>)
                at nl.nn.adapterframework.core.PipeLine.startPipe(PipeLine.java:590)
                at nl.nn.adapterframework.core.PipeLine.start(PipeLine.java:580)
                at nl.nn.adapterframework.core.Adapter$1.run(Adapter.java:817)
                at java.lang.Thread.run(Thread.java:748)
17-Feb-2023 16:15:57.029 INFO [main] org.apache.catalina.core.ApplicationContext.log Successfully created IAF-API with SpringBus [ff-api-bus]
```

I made a commit of my project and pushed it to GitHub. See https://github.com/ibissource/frank-manual/tree/accessMongoDBWithSecrets/src/Frank2MongoDB.

I did not check in file `src/Frank2MongoDB/classes/credentials.properties`. It reads:

```
mongoAccess/username=martijn
mongoAccess/password=zeEenPRAten0134%24
```

You can see on GitHub that my `context.xml` is in the root of the Frank application. The `<Resource>` to access MongoDB reads:

```
        <Resource
                name="mongodb/frank2mongodb"
                type="com.mongodb.MongoClient"
                factory="com.mongodb.client.MongoClientFactory"
                connectionString="mongodb+srv://${mongoAccess/username}:${mongoAccess/password}@cluster0.ktth0z5.mongodb.net/test"
                auth="Container"
                singleton="true"
        />
```

The issue is not that the Frank!Framework cannot resolve the credentials as can be concluded from the following screenshot:

![image](https://user-images.githubusercontent.com/6784732/219700080-f9de7684-4fa7-4215-a60c-2ccb2469c3ca.png)

**Reporter**
Martijn Dirkse

**To Reproduce**
1. Get the Frank application at the URL shown above.
2. Open that directory in VS Code.
3. Run the project with Ant.
4. Check the Tomcat window to see the Java exception.

**Screenshots**

**Configuration**
```
Please provide the configuration of the Pipe or Receiver with the problem.
```

**Input**
```
Please provide an example of the input message. 
Alternatively, provide a Ladybug report.
```

**Environment**
I closed Tomcat. If I restart it, the Frank!Runner may download a new F!F version. I can see I have F!F version 7.9-20230214.195451. I am using the Frank!Runner, commit `67e22ede0f778db6f62e14844629480140982bbd`.

**Additional Environment**
 - I am accessing MongoDB Atlas.

**Additional Context**
This issue arises because I want to document MongoDB. As a first step I was happy to have access, even with hard-coded credentials stored under version control. Now I want to remove these credentials from version control. From a security perspective this may not be important because the accessed database only has example data provided by MongoDB Atlas. We want to show however that other users can use MongoDB in a secure way.

**Comment by mhdirkse** [comment]
This afternoon, @nielsm5, @jacodg and @philipsens helped me with this issue. We tried to put `credentialprovider-7.9-20230214.195451.jar`, `credentialprovider.properties` and `credentials.properties` in the `lib` folder of the Tomcat used by the Frank!Runner. This did not fix the issue.

**Comment by nielsm5** [comment]
Did you configure the credential provider to replace properies in the context.xml?
`org.apache.tomcat.util.digester.PROPERTY_SOURCE=org.apache.tomcat.util.digester.EnvironmentPropertySource,nl.nn.credentialprovider.CredentialProvidingPropertySource`
Needs to be present in the catalina.properties

**Comment by mhdirkse** [comment]
To follow this suggestion using the Frank!Runner, I had to add the following to `build.xml` in my project:

```
<arg value="-Dorg.apache.tomcat.util.digester.PROPERTY_SOURCE=&quot;org.apache.tomcat.util.digester.EnvironmentPropertySource,nl.nn.credentialprovider.CredentialProvidingPropertySource&quot;" />
```

With that change in my project, I could access MongoDB. @jacodg, we checked that we need to add the three mentioned files to Tomcat's `lib` folder.

This result establishes that this is an issue of the Frank!Runner, not the F!F. I close this issue and add an issue to the Frank!Runner.
